src/gui/elem.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+Minetest
+Copyright (C) 2023 v-rob, Vincent Robinson <robinsonvincent89@gmail.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "gui/elem.h"
+
+#include "debug.h"
+#include "log.h"
+#include "gui/manager.h"
+#include "gui/window.h"
+#include "util/serialize.h"
+
+// Include every element header for Elem::create()
+#include "gui/generic_elems.h"
+
+namespace ui
+{
+	std::unique_ptr<Elem> Elem::create(Type type, Window &window, std::string id)
+	{
+		std::unique_ptr<Elem> elem = nullptr;
+
+#define CREATE(name, type)										\
+	case name:													\
+		elem = std::make_unique<type>(window, std::move(id));	\
+		break
+
+		switch (type) {
+			CREATE(ELEM, Elem);
+			CREATE(ROOT, Root);
+		default:
+			return nullptr;
+		}
+
+#undef CREATE
+
+		// It's a pain to call reset() in the constructor of every single
+		// element due to how virtual functions work in C++, so we reset
+		// elements after creating them here.
+		elem->reset();
+		return elem;
+	}
+
+	Elem::Elem(Window &window, std::string id) :
+		m_window(window),
+		m_id(std::move(id)),
+		m_main_box(*this)
+	{}
+
+	void Elem::reset()
+	{
+		m_order = (size_t)-1;
+
+		m_parent = nullptr;
+		m_children.clear();
+
+		m_main_box.reset();
+	}
+
+	void Elem::read(std::istream &is)
+	{
+		u32 set_mask = readU32(is);
+
+		if (testShift(set_mask))
+			readChildren(is);
```

Comment:
Doesn't this approach of serializing all elements in a flat manner and restoring the hierarchy later mean that there can be orphan elements? How do you deal with them?

Reply:
> The `checkTree` function makes sure that all element have a parent unless they are the root of the window.

As far as I can see, `checkTree` only iterates over elements inside the tree / with parents? Am I missing something here?

Reply:
Silly me, I totally thought I checked this somewhere.  I will definitely add this.